test/test-maintenance/BSCValidatorSet.spec.ts

```diff
@@ -0,0 +1,647 @@
+  });
+
+  it('common case 1-22: validator-4 exitMaintenance', async () => {
+    await setSlashIndicator(slashIndicator.address, validatorSet, instances);
+    expect(validatorSet.connect(signers[4]).exitMaintenance()).to.revertedWith(
+      'not in maintenance'
+    );
+  });
+
+  it('common case 1-23: validator-4 enterMaintenance failed!', async () => {
+    expect(validatorSet.connect(signers[4]).enterMaintenance()).to.be.revertedWith(
+      'can not enter Temporary Maintenance'
+    );
+  });
+
+  it('common case 1-24: 24 hours ended, clear all maintainInfo', async () => {
+    expect(await validatorSet.getMaintainingValidators()).to.deep.eq([
+      validators[5],
+      validators[6],
+      validators[8],
+      validators[10],
+    ]);
+    expect(await validatorSet.numOfMaintaining()).to.be.eq(4);
+
+    await waitTx(
+      validatorSet.updateContractAddr(
+        instances[10].address,
+        instances[8].address,
+        instances[3].address,
+        instances[4].address,
+        instances[5].address,
+        instances[0].address,
+        instances[7].address,
+        instances[9].address,
+        instances[6].address,
+        operator.address
+      )
+    );
+
+    await mineBlocks(21 * 50 * maintainSlashScale);
+
+    // do update validators
+    let packageBytes = validatorUpdateRlpEncode(
+      validators.slice(2, 23),
+      validators.slice(2, 23),
+      validators.slice(2, 23)
+    );
+    await waitTx(validatorSet.connect(operator).handleSynPackage(STAKE_CHANNEL_ID, packageBytes));
+
+    // validator-5,6 felony,  validator-8 misdemeanor
+    const expectedValidators = validators
+      .slice(2, 23)
+      .filter((item) => item !== validators[5] && item !== validators[6]);
+    expect(await validatorSet.getValidators()).to.deep.eq(expectedValidators);
+    expect(await validatorSet.numOfMaintaining()).to.be.eq(0);
+
+    for (let i = 2; i < 23; i++) {
+      if (i === 5 || i === 6) {
+        // because of felony, validator-5,6 are not the current validators
+        expect(validatorSet.getCurrentValidatorIndex(validators[i])).to.be.revertedWith(
+          'only current validators'
+        );
+        continue;
+      }
+      const index = await validatorSet.getCurrentValidatorIndex(validators[i]);
+      const validatorExtra = await validatorSet.validatorExtraSet(index);
+
+      expect(validatorExtra.isMaintaining).to.be.eq(false);
+      expect(validatorExtra.enterMaintenanceHeight.toNumber() === 0).to.be.eq(true);
+    }
+    expect(await validatorSet.getMaintainingValidators()).to.deep.eq([]);
+  });
+
+
+  it('common case 1-25: query all view func', async () => {
+    expect(await validatorSet.maxNumOfMaintaining()).to.be.eq(maxNumOfMaintaining)
+    expect(await validatorSet.maintainSlashScale()).to.be.eq(maintainSlashScale)
+    expect(await validatorSet.numOfMaintaining()).to.be.eq(0);
+
+    for (let i = 2; i < 50; i++) {
+      if (i >= 2 && i < 23 && i !== 1 && i !== 5 && i !== 6) {
```

Comment:
`i` started from 2, so I think `(i < 23 && !== 5 && !== 6)` is fine?

Reply:
yes, fixed